crates/red_knot_python_semantic/src/types/context.rs

```diff
@@ -221,3 +201,268 @@ pub(crate) enum InNoTypeCheck {
+            id,
+            severity,
+            lint_id,
+        })
+    }
+
+    /// Create a new lint reporter.
+    ///
+    /// This initializes a new diagnostic using the given message along with
+    /// the ID and severity derived from the `LintMetadata` used to create this
+    /// builder.
+    ///
+    /// The diagnostic can be further mutated via
+    /// `LintReporter::diagnostic`.
+    #[must_use]
+    pub(super) fn build(self, message: impl std::fmt::Display) -> LintReporter<'db, 'ctx> {
+        let diag = Some(Diagnostic::new(self.id, self.severity, message));
+        LintReporter {
+            ctx: self.ctx,
+            diag,
+            lint_id: self.lint_id,
+        }
+    }
+}
+
+/// An abstraction for reporting diagnostics.
+///
+/// Callers can build a reporter via `InferContext::report`.
+///
+/// A reporter encapsulates the logic for determining if a diagnostic *should*
+/// be reported according to the environment, configuration, and any relevant
+/// suppressions. The advantage of this reporter is that callers may avoid
+/// doing extra work to populate a diagnostic if it is known that it would be
+/// otherwise ignored.
+///
+/// The diagnostic is added to the typing context, if appropriate, when this
+/// reporter is dropped.
+///
+/// Callers likely should use `LintReporter` via `InferContext::lint` instead.
+/// This reporter is only intended for use with non-lint diagnostics.
+pub(super) struct DiagnosticReporter<'db, 'ctx> {
+    ctx: &'ctx InferContext<'db>,
+    /// The diagnostic that we want to report.
+    ///
+    /// This is always `Some` until the `Drop` impl.
+    diag: Option<Diagnostic>,
+}
+
+impl DiagnosticReporter<'_, '_> {
+    /// Return a mutable borrow of the diagnostic on this reporter.
+    ///
+    /// Callers may mutate the diagnostic to add new sub-diagnostics
+    /// or annotations.
+    ///
+    /// The diagnostic is added to the typing context, if appropriate,
+    /// when this reporter is dropped.
+    pub(super) fn diagnostic(&mut self) -> &mut Diagnostic {
+        // OK because `self.diag` is only `None` within `Drop`.
+        self.diag.as_mut().unwrap()
+    }
+}
+
+/// Finishes use of this reporter.
+///
+/// This will add the diagnostic to the typing context if appropriate.
+impl Drop for DiagnosticReporter<'_, '_> {
+    fn drop(&mut self) {
+        // The comment below was copied from the original
+        // implementation of diagnostic reporting. The code
+        // has been refactored, but this still kind of looked
+        // relevant, so I've preserved the note. ---AG
+        //
+        // TODO: Don't emit the diagnostic if:
+        // * The enclosing node contains any syntax errors
+        // * The rule is disabled for this file. We probably want to introduce a new query that
+        //   returns a rule selector for a given file that respects the package's settings,
+        //   any global pragma comments in the file, and any per-file-ignores.
+
```

Comment:
I think this fits better into `LintReporter` because it's only there where we have rules.